File: programmers/level2/87946.js
//https://school.programmers.co.kr/learn/courses/30/lessons/87946
//피로도

function solution(k, dungeons) {
  let answer = -1;

  function recursive(result, remain) {
    // console.log('result', result);
    // console.log('remain', remain);

    if (result.length === dungeons.length) {
      // console.log('----result', result);
      let power = k;
      const count = result.reduce((acc, cur) => {
        if (power >= cur[0]) {
          power -= cur[1];
          acc++;
        }
        // console.log('acc', acc);
        return acc;
      }, 0);

      answer = Math.max(answer, count);
      return;
    }

    for (let i = 0; i < remain.length; i++) {
      const rCopy = Array.from(result, ([a, b]) => [a, b]);
      const newResult = [...rCopy, [remain[i][0], remain[i][1]]];
      const rest = remain.filter((_, idx) => idx !== i);
      recursive(newResult, [...rest]);
    }
  }

  for (let i = 0; i < dungeons.length; i++) {
    const dCopy = Array.from(dungeons, ([a, b]) => [a, b]);
    const temp = [[dCopy[i][0], dCopy[i][1]]];
    const rest = dCopy.filter((_, idx) => idx !== i);
    recursive(temp, [...rest]);
  }

  return answer;
}

console.log(
  solution(80, [
    [80, 20],
    [50, 40],
    [30, 10],
  ]),
); //3

File: programmers/python/level2/87946.py
def solution(k, dungeons):
  idx_list = [idx for idx in range(len(dungeons))]
  def shuffle_idx(temp):
    # print('temp',temp)
    if len(temp) == len(dungeons):
      result.append(temp)
      return

    for idx in idx_list:
      if idx not in temp:
        new = [i for i in temp]
        new.append(idx)
        # print('new',new)
        shuffle_idx(new)

  result=[]
  for idx in idx_list:
    # print('-'*50)
    # print('idx',idx)
    shuffle_idx([idx])
    # print('result',result)


  max_cnt = 0
  for temp in result:
    answer = 0
    user_degree=k
    # print('temp',temp)
    for idx in temp:
      # print('dungeon',dungeons[idx])
      min_degree, minus_degree = dungeons[idx]
      # print('user_degree',user_degree)
      if user_degree >= min_degree:
        user_degree -= minus_degree
        answer += 1
      else:
        break

    if answer > max_cnt:
      max_cnt = answer
      # print('max',max_cnt)

  return max_cnt

print(solution(80,	[[80,20],[50,40],[30,10]]))
# 3
